### PR TITLE
Improve OperationTracker to terminate when there are two NOT_FOUND from originating DC

### PR DESCRIPTION
**Pull request by justinlin-linkedin**
NOT_FOUND will be treated as a special error code for OperationTracker.

When there are two NOT_FOUND responses from originating dc, operation tracker would terminate the operation (including GetBlob, GetBlobInfo, DeleteBlob and TtlUpdate). The reason we choose to terminate at two, is because all the mutation operations would require two success responses and we have three replicas. So N + M > R.

### CODE REVIEW

**Comment by justinlin-linkedin** [comment]
This PR is working on the same problem as #1191

File: ambry-router/src/main/java/com.github.ambry.router/DeleteOperation.java [hunk 1]
```diff
@@ -237,80 +251,52 @@ private void cleanupExpiredInflightRequests() {
...
-      if (getPrecedenceLevel(error) < getPrecedenceLevel(resolvedRouterErrorCode)) {
-        resolvedRouterErrorCode = error;
-      }
-    }
-    operationTracker.onResponse(replica,
-        resolvedRouterErrorCode == RouterErrorCode.OperationTimedOut ? TrackedRequestFinalState.TIMED_OUT
-            : TrackedRequestFinalState.FAILURE);
-    if (error != RouterErrorCode.BlobDeleted && error != RouterErrorCode.BlobExpired) {
+  void onErrorResponse(ReplicaId replicaId, RouterException exception) {
+    logger.info("Get exception: " + exception);
```
Comment (jsjtzyy) [inline ambry-router/src/main/java/com.github.ambry.router/DeleteOperation.java]: I think we don't need log here because in most places where `onErrorResponse` is called, there are TRACE logs.

**Comment by jsjtzyy** [comment]
@justinlin-linkedin  could you rebase this PR on top of latest master branch first? (Ensure it uses same version libraries)

File: ambry-router/src/test/java/com.github.ambry.router/GetBlobInfoOperationTest.java [hunk 2]
```diff
@@ -353,6 +353,104 @@ public void testTimeoutAndBlobNotFound() throws Exception {
         tracker.getLatencyHistogram(remoteReplica).getCount());
   }
 
+  /**
+   * Test the case where originating replicas return Blob_Not_found and the rest returns IO_Error.
+   * @throws Exception
+   */
+  @Test
+  public void testIOErrorAndBlobNotFoundOriginatingNotFound() throws Exception {
```
Comment (jsjtzyy) [inline ambry-router/src/test/java/com.github.ambry.router/GetBlobInfoOperationTest.java]: can we make the test name more concise? `testIOErrorAndBlobNotFoundInOriginDc`?

File: ambry-router/src/test/java/com.github.ambry.router/GetBlobInfoOperationTest.java [hunk 3]
```diff
@@ -353,6 +353,104 @@ public void testTimeoutAndBlobNotFound() throws Exception {
...
+    routerConfig = new RouterConfig(new VerifiableProperties(props));
+  }
+
+
+  /**
+   * Test the case where originating replicas return Blob_Not_found and the rest times out.
+   * @throws Exception
+   */
+  @Test
+  public void testTimeoutAndBlobNotFoundOriginatingNotFound() throws Exception {
```
Comment (jsjtzyy) [inline ambry-router/src/test/java/com.github.ambry.router/GetBlobInfoOperationTest.java]: same here for the name.

File: ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java [hunk 4]
```diff
@@ -531,6 +531,54 @@ public void testRequestTimeoutAndBlobNotFound() throws Exception {
         crossColoTracker.getCount());
   }
 
+  /**
+   * Test the case where originating replicas return Blob_Not_found and the rest times out.
+   * @throws Exception
+   */
+  @Test
+  public void testTimeoutAndBlobNotFoundOriginatingNotFound() throws Exception {
```
Comment (jsjtzyy) [inline ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java]: same here

File: ambry-router/src/test/java/com.github.ambry.router/OperationTrackerTest.java [hunk 5]
```diff
@@ -431,6 +431,79 @@ public void replicasOrderTestOriginatingDcOnly() {
     assertTrue("Operation should be done", ot.isDone());
   }
 
+  /**
+   * Test the case when NotFound Error should be disabled since the cross colo is disabled.
+   */
+  @Test
+  public void originatingDcNotFoundColoDisableTest() {
```
Comment (jsjtzyy) [inline ambry-router/src/test/java/com.github.ambry.router/OperationTrackerTest.java]: `blobNotFoundInOriginDcAndCrossColoDisabledTest`?

File: ambry-router/src/main/java/com.github.ambry.router/DeleteOperation.java [hunk 6]
```diff
@@ -192,15 +192,29 @@ void handleResponse(ResponseInfo responseInfo, DeleteResponse deleteResponse) {
...
+          if (getError == ServerErrorCode.No_Error || getError == ServerErrorCode.Blob_Deleted) {
+            operationTracker.onResponse(replica, TrackedRequestFinalState.SUCCESS);
+            if (RouterUtils.isRemoteReplica(routerConfig, replica)) {
+              logger.trace("Cross colo request successful for remote replica {} in {} ", replica.getDataNodeId(),
+                  replica.getDataNodeId().getDatacenterName());
+              routerMetrics.crossColoSuccessCount.inc();
+            }
+          } else {
+            logger.trace("Replica {} returned an error {} for a delete request with response correlationId : {} ",
+                replica.getDataNodeId(), getError, deleteRequest.getCorrelationId());
```
[the remark this answers is not in the document]
Comment (justinlin-linkedin) [reply ambry-router/src/main/java/com.github.ambry.router/DeleteOperation.java]: are you saying I should just use replica, instead of replica.getDataNodeID()?

File: ambry-router/src/main/java/com.github.ambry.router/GetOperation.java [hunk 7]
```diff
@@ -184,7 +184,7 @@ void setOperationException(Exception exception) {
    * @param routerErrorCode The {@link RouterErrorCode} for which to get its precedence level.
    * @return The precedence level of the {@link RouterErrorCode}.
    */
-  private int getPrecedenceLevel(RouterErrorCode routerErrorCode) {
+  protected int getPrecedenceLevel(RouterErrorCode routerErrorCode) {
```
[the remark this answers is not in the document]
Comment (justinlin-linkedin) [reply ambry-router/src/main/java/com.github.ambry.router/GetOperation.java]: this function should be visible to it's subclasses, GetBlobInfoOperation and GetBlobOperation. Protected should be a better modifier here.

File: ambry-router/src/main/java/com.github.ambry.router/TtlUpdateOperation.java [hunk 8]
```diff
@@ -313,14 +302,32 @@ private void updateOperationState(ReplicaId replica, RouterErrorCode error) {
...
 
+  /**
+   * Set the exception associated with this operation.
+   * First, if current operationException is null, directly set operationException as exception;
+   * Second, if operationException exists, compare ErrorCodes of exception and existing operation Exception depending
+   * on precedence level. An ErrorCode with a smaller precedence level overrides an ErrorCode with a larger precedence
+   * level. Update the operationException if necessary.
+   * @param exception the {@link RouterException} to possibly set.
+   */
+  void setOperationException(Exception exception) {
```
[the remark this answers is not in the document]
Comment (justinlin-linkedin) [reply ambry-router/src/main/java/com.github.ambry.router/TtlUpdateOperation.java]: It's not necessarily a `RouterException`, at least in GetOperation. When reading bytes from tcp channel can result in non-RouterException.

[the remark this answers is not in the document]
Comment (jsjtzyy) [reply ambry-router/src/main/java/com.github.ambry.router/TtlUpdateOperation.java]: True, but the method in DeleteOperation and TtlUpdateOperation is always taking RouterException. I feel like `operationException.compareAndSet(null, exception);` in `else` block never gets invoked.

[the remark this answers is not in the document]
Comment (justinlin-linkedin) [reply ambry-router/src/main/java/com.github.ambry.router/TtlUpdateOperation.java]: right, will update

File: ambry-router/src/main/java/com.github.ambry.router/TtlUpdateOperation.java [hunk 9]
```diff
@@ -51,7 +52,7 @@
   // The operation tracker that tracks the state of this operation.
   private final OperationTracker operationTracker;
   // A map used to find inflight requests using a correlation id.
-  private final Map<Integer, TtlUpdateRequestInfo> ttlUpdateRequestInfos = new HashMap<>();
+  private final Map<Integer, TtlUpdateRequestInfo> ttlUpdateRequestInfos = new TreeMap<>();
```
Comment (dharju) [inline ambry-router/src/main/java/com.github.ambry.router/TtlUpdateOperation.java]: why this change?

File: ambry-router/src/main/java/com.github.ambry.router/DeleteOperation.java [hunk 10]
```diff
@@ -96,7 +95,7 @@
     this.callback = callback;
     this.time = time;
     this.deletionTimeMs = time.milliseconds();
-    this.deleteRequestInfos = new HashMap<Integer, DeleteRequestInfo>();
+    this.deleteRequestInfos = new TreeMap<Integer, DeleteRequestInfo>();
```
Comment (dharju) [inline ambry-router/src/main/java/com.github.ambry.router/DeleteOperation.java]: Why use treemap here?

File: ambry-router/src/main/java/com.github.ambry.router/DeleteOperation.java [hunk 11]
```diff
@@ -319,14 +301,26 @@ private void updateOperationState(ReplicaId replica, RouterErrorCode error) {
...
 
+  /**
+   * Set the exception associated with this operation.
+   * If operationException exists, compare ErrorCodes of exception and existing operation Exception depending
+   * on precedence level. An ErrorCode with a smaller precedence level overrides an ErrorCode with a larger precedence
+   * level. Update the operationException if necessary.
+   * @param exception the {@link RouterException} to possibly set.
+   */
+  void setOperationException(RouterException exception) {
+    RouterUtils.replaceOperationException(operationException, (RouterException) exception, this::getPrecedenceLevel);
```
Comment (dharju) [inline ambry-router/src/main/java/com.github.ambry.router/DeleteOperation.java]: You don't need to cast here since you're accepting a `RouterException`

File: ambry-router/src/main/java/com.github.ambry.router/DeleteOperation.java (hunk 10 above)
Comment (justinlin-linkedin) [reply ambry-router/src/main/java/com.github.ambry.router/DeleteOperation.java]: In fetchRequests function, this map will be iterated based on the order of it's key. So I change it to a sorted tree.

File: ambry-router/src/main/java/com.github.ambry.router/TtlUpdateOperation.java (hunk 9 above)
Comment (justinlin-linkedin) [reply ambry-router/src/main/java/com.github.ambry.router/TtlUpdateOperation.java]: In fetchRequests function, this map will be iterated based on the order of it's key. So I change it to a sorted tree.

File: ambry-router/src/main/java/com.github.ambry.router/DeleteOperation.java (hunk 11 above)
Comment (justinlin-linkedin) [reply ambry-router/src/main/java/com.github.ambry.router/DeleteOperation.java]: updated.

File: ambry-router/src/main/java/com.github.ambry.router/TtlUpdateOperation.java [hunk 12]
```diff
@@ -313,14 +302,27 @@ private void updateOperationState(ReplicaId replica, RouterErrorCode error) {
...
 
+  /**
+   * Set the exception associated with this operation.
+   * If operationException exists, compare ErrorCodes of exception and existing operation Exception depending
+   * on precedence level. An ErrorCode with a smaller precedence level overrides an ErrorCode with a larger precedence
+   * level. Update the operationException if necessary.
+   * @param exception the {@link RouterException} to possibly set.
+   */
+  void setOperationException(Exception exception) {
+    RouterUtils.replaceOperationException(operationException, (RouterException) exception, this::getPrecedenceLevel);
```
[the remark this answers is not in the document]
Comment (justinlin-linkedin) [reply ambry-router/src/main/java/com.github.ambry.router/TtlUpdateOperation.java]: you are right, updated.

File: ambry-router/src/test/java/com.github.ambry.router/DeleteManagerTest.java [hunk 13]
```diff
@@ -255,6 +262,13 @@ public void routerErrorCodeResolutionTest() throws Exception {
     codesToSetAndTest.put(ServerErrorCode.Replica_Unavailable, RouterErrorCode.AmbryUnavailable);
     codesToSetAndTest.put(ServerErrorCode.Partition_Unknown, RouterErrorCode.UnexpectedInternalError);
     doRouterErrorCodeResolutionTest(codesToSetAndTest);
+
```
Comment (dharju) [inline ambry-router/src/test/java/com.github.ambry.router/DeleteManagerTest.java]: You need this code block here?  Looks like it was accidentally copied here from above

File: ambry-router/src/main/java/com.github.ambry.router/TtlUpdateOperation.java [hunk 14]
```diff
@@ -313,14 +302,27 @@ private void updateOperationState(ReplicaId replica, RouterErrorCode error) {
   private void checkAndMaybeComplete() {
     // operationCompleted is true if Blob_Authorization_Failure was received.
     if (operationTracker.isDone() || operationCompleted) {
-      if (!operationTracker.hasSucceeded()) {
-        setOperationException(
-            new RouterException("The TtlUpdateOperation could not be completed.", resolvedRouterErrorCode));
+      if (operationTracker.hasSucceeded()) {
+        operationException.set(null);
```
Comment (dharju) [inline ambry-router/src/main/java/com.github.ambry.router/TtlUpdateOperation.java]: It doesn't default to null?  It's possible to not be null at this point?

File: ambry-router/src/main/java/com.github.ambry.router/DeleteOperation.java (hunk 10 above)
Comment (dharju) [reply ambry-router/src/main/java/com.github.ambry.router/DeleteOperation.java]: how does that affect functionality?

File: ambry-api/src/main/java/com.github.ambry/config/RouterConfig.java [hunk 15]
```diff
@@ -298,6 +298,15 @@
   @Default("1000")
   public final long routerOperationTrackerMinDataPointsRequired;
 
+  /**
```
Comment (dharju) [inline ambry-api/src/main/java/com.github.ambry/config/RouterConfig.java]: Did you run the code formatter on all the files?

File: ambry-router/src/test/java/com.github.ambry.router/DeleteManagerTest.java (hunk 13 above)
Comment (justinlin-linkedin) [reply ambry-router/src/test/java/com.github.ambry.router/DeleteManagerTest.java]: Just realize this is an old change. updated.

File: ambry-router/src/main/java/com.github.ambry.router/TtlUpdateOperation.java (hunk 14 above)
Comment (justinlin-linkedin) [reply ambry-router/src/main/java/com.github.ambry.router/TtlUpdateOperation.java]: it's defaulted to be null, but it's also possible to be not null at this point. we will send three requests to ambry-server and every error response for those requests will set the operationException to something not null.

File: ambry-api/src/main/java/com.github.ambry/config/RouterConfig.java (hunk 15 above)
Comment (justinlin-linkedin) [reply ambry-api/src/main/java/com.github.ambry/config/RouterConfig.java]: updated.

File: ambry-router/src/main/java/com.github.ambry.router/DeleteOperation.java (hunk 10 above)
Comment (justinlin-linkedin) [reply ambry-router/src/main/java/com.github.ambry.router/DeleteOperation.java]: Since DeleteRequestInfo has a creationtTime, and the bigger the key is in the map, the latter the creationTime is in the request. We can rely on this correlation while trying to find all the timed out requests. As long as we hit a request that is not timed out, we can be sure that all the requests following that is not timed out.
